File: codes/ruby/chapter06/answer_A40.rb
# 入力
N = gets.chomp.to_i
A = gets.chomp.split.map(&:to_i)

# 個数を数える
cnt = [0] * 101
Answer = 0
A.each do |a|
  cnt[a] += 1
end

# 答えを求める
(1..100).each do |i|
  Answer += cnt[i] * (cnt[i] - 1) * (cnt[i] - 2) / 6
end

puts Answer

File: codes/go/chapter06/answer_A40.go
package main

import (
	"fmt"
)

func main() {
	var N int
	fmt.Scan(&N)
	A := make([]int, N+1)
	for i := 1; i <= N; i++ {
		fmt.Scan(&A[i])
	}

	cnt := make([]int64, 101)
	for i := 1; i <= 100; i++ {
		cnt[i] = 0
	}
	for i := 1; i <= N; i++ {
		cnt[A[i]] += 1
	}

	var Answer int64 = 0
	for i := 1; i <= 100; i++ {
		Answer += cnt[i] * (cnt[i] - 1) * (cnt[i] - 2) / 6
	}
	fmt.Println(Answer)
}
